std/machines/small_field/pointer_arith.asm

```diff
@@ -0,0 +1,35 @@
+/// Creates constraints that increments a pointer by one full 32-bit word.
+///
+/// The pointers are given in 2 16-bit limbs. This function return the polynomials constraining
+/// (int(post_high) * 0x10000 + int(post_low)) to (int(pre_high) * 0x10000 + int(pre_low) + 4).
+///
+/// The returned constraints are not automatically added to the constraints set. This is so the
+/// caller will be able to latch-disable them in rows they aren't needed.
+///
+/// This constr function introduces 2 new witness column and some helper constraints.
+///
+/// Preconditions: pre_low must be 16-bit aligned and in range [0, 0xfffc].
+let word_increment_ptr: expr, expr, expr, expr -> Constr[] = constr |pre_high, pre_low, post_high, post_low| {
+    // How far away from overflowing the low limb is:
+    let low_diff = pre_low - (0x10000 - 4);
+    // Is one if low limb is about to overflow:
```

Comment:
I still don't understand what "about to overflow" means and what the two constraints in lines 20 and 24 do exactly.

Reply:
"about to overflow" in a sense that in the next step, the low limb will be set to zero and the carry will be added to the high limb.

Reply:
About lines 20;24-26, it can be seen as an `is_not_zero` function (could be extracted, if useful somewhere else):
```
let is_not_zero = constr |value| {
    // Is set to the inverse of "value" if it is not zero, otherwise undefined.
    let inverse;

    // Ensures that inverse * value is either 0 or 1:
    (inverse * value - 1) * value = 0;

    inverse * value
}
```

Reply:
I placed this is_not_zero in utils, and changed is_zero implementation to use it (because it is simpler). Now I am running the tests to see if it will break anything, since it returns a degree 2 expression.

Reply:
Renamed `low_overflow` to `carry`, so its role is more recognizable.